.build/docker/run-tests.sh

```diff
@@ -0,0 +1,259 @@
+inner_split_first=$(( ( $(echo $split_chunk | cut -d"/" -f1 ) * ${docker_runs} ) - ( ${docker_runs} - 1 ) ))
+docker_cpus=$(echo "scale=2; ${cores} / ( ${jenkins_executors} * ${docker_runs} )" | bc)
+
+# hack: long-test does not handle limited CPUs
+if [ "${target}" == "long-test" ] ; then
+    docker_flags="-m 5g --memory-swap 5g"
+elif [[ "${target}" =~ dtest* ]] ; then
+    docker_flags="--cpus=${docker_cpus} -m 15g --memory-swap 15g"
+else
+    docker_flags="--cpus=${docker_cpus} -m 5g --memory-swap 5g"
+fi
+docker_flags="${docker_flags} --env-file build/env.list -d --rm"
+
+# make sure build_dir is good
+mkdir -p ${cassandra_dir}/build/tmp || true
+mkdir -p ${cassandra_dir}/build/test/logs || true
+chmod -R ag+rwx ${cassandra_dir}/build
+
+# FIXME python
+python_version="python3.7"
+cython="no"
+
+# the docker container's env
+touch build/env.list
+cat > build/env.list <<EOF
+TEST_SCRIPT=${test_script}
+JAVA_VERSION=${java_version}
+# FIXME python
+PYTHON_VERSION=${python_version}
+CYTHON_ENABLED=${cython}
+ANT_OPTS="-Dtesttag.extra=.arch=$(arch).${python_version}.cython=${cython}"
+EOF
+
+declare -a docker_ids
+declare -a process_ids
+declare -a statuses
+
+for i in `seq 1 ${docker_runs}` ; do
+    inner_split=$(( ${inner_split_first} + ( $i - 1 ) ))
+
+    container_name="cassandra_${dockerfile/.docker/}_${target}_jdk${java_version/./-}_arch-$(arch)_${python_version/./-}_cython-${cython}_${inner_split}_${inner_splits}__$(LC_ALL=C tr -dc A-Za-z0-9 </dev/urandom | head -c 6 ; echo '')"
+
+    # Docker commands:
+    #  chown the existing cassandra/build directory so we can tmpfs over it (to ensure it is not touched)
+    #  set java to java_version
+    #  execute the run_script
+
+    docker_command="sudo chown -R cassandra \${CASSANDRA_DIR}/build \${DIST_DIR} ; \
+                source \${CASSANDRA_DIR}/.build/docker/_set_java.sh ${java_version} ; \
+                \${CASSANDRA_DIR}/.build/docker/_docker_init_tests.sh ${target} ${inner_split}/${inner_splits} ; exit \$?"
+
+    # start the container
+    docker_id=$(docker run --name ${container_name} ${docker_flags} ${docker_mounts} ${image_name} sleep inf)
+
+    echo "Running container ${container_name} ${docker_id}"
+
+    docker exec --user root ${container_name} bash -c "\${CASSANDRA_DIR}/.build/docker/_create_user.sh cassandra $(id -u) $(id -g)"
+
+    # capture logs and pid for container
+    docker exec --user cassandra ${container_name} dumb-init bash -ilc -c "${docker_command}" > build/test/logs/docker_attach_${i}.log &
+    #docker attach --no-stdin ${docker_id} > build/test/logs/docker_attach_${i}.log &
+    process_ids+=( $! )
+    docker_ids+=( ${docker_id} )
+done
+
+exit_result=0
+i=0
+for process_id in "${process_ids[@]}" ; do
+    # wait for each container to complete
+    docker_id=${docker_ids[$i]}
+    inner_split=$(( $inner_split_first + $i ))
+    cat build/test/logs/docker_attach_$(( $i + 1 )).log
+    tail -F build/test/logs/docker_attach_$(( $i + 1 )).log &
```

Comment:
Wasn't clear to me why we're tailing the process here; is the hope to catch things that hang and have visibility into them? If so, a comment would help clarify.

Reply:
these docker loops (inner splits) may go altogether…